### PR TITLE
DROID-62, DROID-71: MeatNet Scanning & Advertising Data

### PR DESCRIPTION
**Pull request by miwright2**
- Timer advertising data is disabled in code, see this code in `ProbeManager`

```
    private fun debuggingWithStaticLink(device: ProbeBleDeviceBase): Boolean {
        return when(device) {
            is ProbeBleDevice -> true
            is RepeatedProbeBleDevice -> false
            else -> false
        }
    }
```

### CODE REVIEW

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt [hunk 1]
```diff
@@ -159,85 +229,46 @@ internal open class BleDevice (
...
 
-        // the probe continues to advertise even while a BLE connection is
-        // established.  determine if the device is currently advertising as
-        // connectable or not.
-        val advertisingState = when(newAdvertisingData.isConnectable) {
-            true -> DeviceConnectionState.ADVERTISING_CONNECTABLE
-            else -> DeviceConnectionState.ADVERTISING_NOT_CONNECTABLE
-        }
+                // if advertising packet has same mac address and same probe serial number
+                mac == it.mac && it.probeSerialNumber == serialNumber
```
Comment (nick-sasquatch) [inline combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt:237]: Matt and I talked offline about moving some of this probe-specific functionality closer to where it's used.

**Review by nick-sasquatch — CHANGES_REQUESTED** [review]
Few small items and a bit of a refactor we discussed earlier. All in all though, this looks great--added a bunch of functionality while also managing to clean up the interface. Nice work.

File: combustion-android-ble/src/main/java/inc/combustion/framework/service/NetworkEvent.kt [hunk 2]
```diff
@@ -25,8 +25,13 @@
...
+package inc.combustion.framework.service
 
-package inc.combustion.framework.ble.device;
-
-public class SimulatedProbeBleDevice {
-}
+/**
+ * Enumerates the current network state
+ */
+data class NetworkState(
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/service/NetworkEvent.kt:33]: oops, yes!

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt [hunk 3]
```diff
@@ -72,16 +88,18 @@ internal class NetworkManager(
...
                         BluetoothAdapter.STATE_OFF -> {
-                            mutableDiscoveredProbesFlow.tryEmit(DeviceDiscoveredEvent.ScanningOff)
-                            mutableDiscoveredProbesFlow.tryEmit(DeviceDiscoveredEvent.BluetoothOff)
-                        }
-                        BluetoothAdapter.STATE_TURNING_OFF -> {
-                            LegacyDeviceScanner.stopProbeScanning()
+                            DeviceScanner.stop()
+                            networkState.value = networkState.value.copy(
+                                scanningOn = false,
+                                bluetoothOn = false
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt:94]: good call -- i'll assign it to the default state `NetworkState()`

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt [hunk 4]
```diff
@@ -96,24 +114,56 @@ internal class NetworkManager(
...
     init {
         jobManager.addJob(owner.lifecycleScope.launch {
-            // TODO: Need to start the device scanner
+            collectAdvertisingData()
         })
 
+        // if MeatNet is enabled,then automatically start and stop scanning state
+        // based on current bluetooth state.  this code cannot handle application permissions
+        // in that case, the client app needs to start scanning once bluetooth permissions are allowed.
+        if(SETTINGS.meatNetEnabled) {
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt:135]: i was able to tease it apart in `CombustionService` and have it be read-only in `NetworkManager` and below.  it needs to be a `lateinit var` in `CombustionService` with how the service is constructed versus how it is started.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt [hunk 5]
```diff
@@ -250,50 +306,76 @@ internal class NetworkManager(
...
+                    DeviceHolder.ProbeHolder(
+                        ProbeBleDevice(advertisement.mac, owner, advertisement, adapter)
+                    )
+                }
+                CombustionProductType.DISPLAY -> {
+                    DeviceHolder.RepeaterHolder(
+                        UartBleDevice(advertisement.mac, owner, advertisement, adapter)
+                    )
+                }
+                CombustionProductType.CHARGER -> {
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt:328]: thanks!

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt [hunk 6]
```diff
@@ -250,50 +306,76 @@ internal class NetworkManager(
...
-                        // if not, then create a new probe manager for it.serial
-                        // create a RepeatedProbeBleDevice for it.serial
-                        // add it to the new Probe manager
-
-                        // if we created a new ProbeBleDeive
-                        /*
-                        _discoveredProbesFlow.emit(
-                            DeviceDiscoveredEvent.DeviceDiscovered(it.serialNumber)
+                if(manageMeatNetDevice(it)) {
+                    if(it.probeSerialNumber != "0") {
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt:372]: i'm guessing something else, but i'll define it.

[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt:372]: and add a comment to the define.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt [hunk 7]
```diff
@@ -135,110 +187,114 @@ internal class NetworkManager(
...
     }
 
     fun clearDevices() {
-        /*
-        _probes.forEach { (_, probe) -> probe.finish() }
-        _probes.clear()
-         */
-        mutableDiscoveredProbesFlow.tryEmit(DeviceDiscoveredEvent.DevicesCleared)
-        TODO()
+        probeManagers.forEach { (_, probe) -> probe.finish() }
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/NetworkManager.kt:288]: I don't think so.   `finish()` (once implemented) should synchronously cancel all async jobs using `JobManager`.  I think the garbage collector and `Job` are able to dispose of any resources in the coroutine scope once it is done being canceled.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/scanning/DeviceScanner.kt [hunk 8]
```diff
@@ -48,8 +48,7 @@ import java.util.concurrent.atomic.AtomicBoolean
 
 internal class DeviceScanner private constructor() {
     companion object {
-        // TODO: This is the wrong UUID.
-        private const val NORDIC_DFU_SERVICE_UUID_STRING = "0000180A-0000-1000-8000-0080BEEFBEEF"
+        private const val NORDIC_DFU_SERVICE_UUID_STRING = "0000FE59-0000-1000-8000-00805F9B34FB"
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/scanning/DeviceScanner.kt:51]: no problem

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/scanning/BaseAdvertisingData.kt [hunk 9]
```diff
@@ -133,11 +133,11 @@ open class BaseAdvertisingData(
                     base
                 }
                 CombustionProductType.PROBE -> {
```
[the remark this answers is not in the document]
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/scanning/BaseAdvertisingData.kt:135]: refactored to remove the `else` so the build will break when we add a new entry to `CombustionProductType`.  That syntax sugar above is nice.  Thanks again.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt [hunk 10]
```diff
@@ -101,14 +100,16 @@ internal open class BleDevice (
...
 
-    private val advertisementForProbe = hashMapOf<String, CombustionAdvertisingData>()
     private val remoteRssi = AtomicInteger(0)
     private val connectionMonitor = IdleMonitor()
 
     val rssi get() = remoteRssi.get()
     var connectionState = DeviceConnectionState.OUT_OF_RANGE
     val isConnected = AtomicBoolean(false)
 
+    /*
```
Comment (nick-sasquatch) [inline combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt:110]: Please remove this commented-out code if it's not going to be used again.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt [hunk 11]
```diff
@@ -229,6 +231,45 @@ internal open class BleDevice (
         })
     }
 
+    fun observeAdvertisingPackets(filter: (advertisement: CombustionAdvertisingData) -> Boolean, callback: (suspend (advertisement: CombustionAdvertisingData) -> Unit)? = null) {
+        jobManager.addJob(owner.lifecycleScope.launch(Dispatchers.IO) {
+            DeviceScanner.advertisements.filter {
+
+                // if advertising packet has same mac address and same probe serial number
```
Comment (nick-sasquatch) [inline combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt:238]: This comment looks like it's out of date.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt [hunk 12]
```diff
@@ -229,6 +231,45 @@ internal open class BleDevice (
...
+                }
+
+                callback?.let { clientCallback ->
+                    clientCallback(it)
+                }
+            }
+        })
+    }
+
+    /*
```
Comment (nick-sasquatch) [inline combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt:272]: Please remove this commented-out code if it's not going to be used again.

**Review by nick-sasquatch — APPROVED** [review]
Appreciate you going through and sorting out these last details. This all looks good to me, and barring the problem you're seeing in the prod app being in the framework, I think this is ready to go. I'm going to rebase off of it for the DFU work. I requested a couple commented-out blocks be removed--if you're tired of dealing with this branch and want to be done with it, I can pick those up in a subsequent PR.

Nice job, Matt.

File: combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt (hunk 10 above)
Comment (miwright2) [reply combustion-android-ble/src/main/java/inc/combustion/framework/ble/device/BleDevice.kt:110]: Sorry about that -- got it cleaned up.

**Pull request merged**